<a href="https://colab.research.google.com/github/Toulouse08/IS-170-Machine-Learning/blob/Association-Rules/Top20Tvshows_170.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules


In [ ]:
df=pd.read_csv('Top20Tvshows.csv')
df.head()

Planet Earth II               Documentary
0      Breaking Bad    Crime, Drama, Thriller
1      Planet Earth               Documentary
2  Band of Brothers       Drama, History, War
3         Chernobyl  Drama, History, Thriller
4          The Wire    Crime, Drama, Thriller

In [ ]:
#Since it's missing a header we'll create one, the new column header 'Products' 
df=pd.read_csv('Top20Tvshows.csv', names=['tags'])
df.head() 


tags
Planet Earth II                Documentary
Breaking Bad        Crime, Drama, Thriller
Planet Earth                   Documentary
Band of Brothers       Drama, History, War
Chernobyl         Drama, History, Thriller

In [ ]:
df.shape

(20, 1)

In [ ]:
#the following code converts a pandas dataframe into a list of lists.
data=list(df["tags"].apply(lambda x:x.split(",")))
data

[['Documentary'],
 ['Crime', ' Drama', ' Thriller'],
 ['Documentary'],
 ['Drama', ' History', ' War'],
 ['Drama', ' History', ' Thriller'],
 ['Crime', ' Drama', ' Thriller'],
 ['Animation', ' Action', ' Adventure', ' Family', ' Fantasy', ' Mystery'],
 ['Documentary'],
 ['Crime', ' Drama'],
 ['Documentary'],
 ['Documentary'],
 ['Documentary'],
 ['Action', ' Adventure', ' Drama', ' Fantasy'],
 ['Documentary', ' History', ' War'],
 ['Animation', ' Adventure', ' Comedy', ' Sci-fi'],
 ['Animation',
  ' Action',
  ' Adventure',
  ' Comedy',
  ' Drama',
  ' Fantasy',
  ' Sci-fi'],
 ['Documentary', ' Biography', ' History', ' Sport'],
 ['Documentary'],
 ['Drama', ' Fantasy', ' Horror', ' Mystery', ' Sci-fi', ' Thriller'],
 ['Crime', ' Drama', ' Mystery', ' Thriller']]

In [ ]:
#Let's transform the list with one-hot encoding (make sure you understand the one-hot encoding concept first)
#Each column represents an item and each row represent a record or a transaction. 
#Initially, if that particular item occurs then it's labeled TRUE, if not then FALSE.
#After that, TRUE is converted to 1 and FALSE to 0 because computer loves numeric values.

from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Turn False & True to 0 & 1 for easy readability
df=df.replace(False,0)
df=df.replace(True,1)

df
                

Action   Adventure   Biography   Comedy   Drama   Family   Fantasy  \
0         0           0           0        0       0        0         0   
1         0           0           0        0       1        0         0   
2         0           0           0        0       0        0         0   
3         0           0           0        0       0        0         0   
4         0           0           0        0       0        0         0   
5         0           0           0        0       1        0         0   
6         1           1           0        0       0        1         1   
7         0           0           0        0       0        0         0   
8         0           0           0        0       1        0         0   
9         0           0           0        0       0        0         0   
10        0           0           0        0       0        0         0   
11        0           0           0        0       0        0         0   
12        0           1           0        0       1        0         1   
13        0           0           0        0       0        0         0   
14        0           1           0        1       0        0         0   
15        1           1           0        1       1        0         1   
16        0           0           1        0       0        0         0   
17        0           0           0        0       0        0         0   
18        0           0           0        0       0        0         1   
19        0           0           0        0       1        0         0   

     History   Horror   Mystery   Sci-fi   Sport   Thriller   War  Action  \
0          0        0         0        0       0          0     0       0   
1          0        0         0        0       0          1     0       0   
2          0        0         0        0       0          0     0       0   
3          1        0         0        0       0          0     1       0   
4          1        0         0        0       0          1     0       0   
5          0        0         0        0       0          1     0       0   
6          0        0         1        0       0          0     0       0   
7          0        0         0        0       0          0     0       0   
8          0        0         0        0       0          0     0       0   
9          0        0         0        0       0          0     0       0   
10         0        0         0        0       0          0     0       0   
11         0        0         0        0       0          0     0       0   
12         0        0         0        0       0          0     0       1   
13         1        0         0        0       0          0     1       0   
14         0        0         0        1       0          0     0       0   
15         0        0         0        1       0          0     0       0   
16         1        0         0        0       1          0     0       0   
17         0        0         0        0       0          0     0       0   
18         0        1         1        1       0          1     0       0   
19         0        0         1        0       0          1     0       0   

    Animation  Crime  Documentary  Drama  
0           0      0            1      0  
1           0      1            0      0  
2           0      0            1      0  
3           0      0            0      1  
4           0      0            0      1  
5           0      1            0      0  
6           1      0            0      0  
7           0      0            1      0  
8           0      1            0      0  
9           0      0            1      0  
10          0      0            1      0  
11          0      0            1      0  
12          0      0            0      0  
13          0      0            1      0  
14          1      0            0      0  
15          1      0            0      0  
16          0      0            1      0  
17          0      0            1    

In [ ]:
#We'll try 0.2 for the SUPPORT threshold value (but you can set it to any value you want).
# The parameter (use_colnames=True) is used to convert the integer back to its respective name.

df=apriori(df,min_support=0.2,use_colnames=True)
df

support         itemsets
0     0.20     ( Adventure)
1     0.30         ( Drama)
2     0.20       ( Fantasy)
3     0.20       ( History)
4     0.25      ( Thriller)
5     0.20          (Crime)
6     0.45    (Documentary)
7     0.20  ( Drama, Crime)

With the SUPPORT values above, you can try and match your own manual calculation. You notice how some are grouped with two. ALL items have been tested alone and in groups, but only these ones with the SUPPORT threshold 0.2 or higher values (i.e. other groups didn't make 0.2 therefore not shown).

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents consequents  antecedent support  consequent support  support  \
0    ( Drama)     (Crime)                 0.3                 0.2      0.2   
1     (Crime)    ( Drama)                 0.2                 0.3      0.2   

   confidence      lift  leverage  conviction  
0    0.666667  3.333333      0.14         2.4  
1    1.000000  3.333333      0.14         inf

From above result, we say that the support is 0.2 calculated by dividing the number of tags containing Drama and Crime by the total number of tags.  The confidence level is 0.67 shows that out of all the tags that contain Drama 0.67 contain Crime too.  The lift 3.33 tells us that Drama is 3.33 times more likely to be present on Tv Shows with the Crime tag.

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar


antecedents consequents  antecedent support  consequent support  support  \
0    ( Drama)     (Crime)                 0.3                 0.2      0.2   
1     (Crime)    ( Drama)                 0.2                 0.3      0.2   

   confidence      lift  leverage  conviction  
0    0.666667  3.333333      0.14         2.4  
1    1.000000  3.333333      0.14         inf

From above result, let's take a look at the first row items: Drama and Crime. 
The two items have the support of 0.2, where Drama is 0.3 and Crime is 0.2. 
confidence is 0.67, this implies that if a Tv Show had Drama then there is 67% Crime is also a tag used. There are also positive correlations as we can see from the positive values of lift, leverage, and conviction.

For any show that will be a success it is seen through ML that Drama and Crime are genre that will keep audience invested in the story and product.